lib/storage/index.js

```diff
@@ -0,0 +1,266 @@
+import _ from 'underscore';
+import PlatformStorage from './platforms';
+import MemoryOnlyProvider from './providers/MemoryOnlyProvider';
+import InstanceSync from './InstanceSync';
+import * as Logger from '../Logger';
+
+let storageProviders = [PlatformStorage, MemoryOnlyProvider];
+let finishInitalization;
+const initPromise = new Promise((resolve) => {
+    finishInitalization = resolve;
+});
+let shouldKeepInstancesSync = false;
+
+/**
+ * Runs an operation consecutively on all storage providers
+ * @param {String} operationName
+ * @param {Array<any>} params - the parameters to pass to the operation
+ * @return {Promise<void>}
+ */
+function runConsecutively(operationName, ...params) {
+    const runOperation = (providerIndex) => {
+        const provider = storageProviders[providerIndex];
+
+        return new Promise((resolve, reject) => {
+            provider[operationName](...params)
+                .then(resolve)
+                .catch((error) => {
+                    if (providerIndex + 1 >= storageProviders.length) {
+                        reject(error);
+                        return;
+                    }
+
+                    runOperation(providerIndex + 1)
```

Comment:
Will this double the memory footprint of the app? It looks like we will have values stored in both the Onyx cache + "memory" provider.

Reply:
I am pretty sure it will double the memory footprint. Which makes sense because this is replacing disc storage with an in-memory storage.

I feel like it's probably OK until we identify that the doubled memory footprint is an actual problem.

Ideally, it would be nice that the Onyx cache and this in memory provider are one-in-the-same, but it is kind of out of scope for this and something we can look at doing in the future.

You can see I commented about the same thing in the GH and you can read Chris's response to it: https://github.com/Expensify/App/issues/29403#issuecomment-1870409516

Reply:
Yes, but only in case the main storage provider (IDBKeyVal/SQLite) fails to initialize.

As described in https://github.com/Expensify/App/issues/29403#issuecomment-1870222315 i planned on completely removing `OnyxCache` in favor of `MemoryOnlyProvider` in a follow-up PR, since it's essentially doing the same thing.

For now, the app would only have a double memory footprint in case of the storage failing.

Is that an approach we can go with?

Reply:
Did we consider the suggestion [here](https://github.com/Expensify/App/issues/29403#issuecomment-1821937151)?

>Having a "provider" that just returns resolved promises for all the methods

Reply:
>completely removing OnyxCache in favor of MemoryOnlyProvider in a follow-up PR, since it's essentially doing the same thing

Just my opinion, but sounds like more work for unclear gain.

Reply:
> Did we consider the suggestion [here](https://github.com/Expensify/App/issues/29403#issuecomment-1821937151)?
> 
> > Having a "provider" that just returns resolved promises for all the methods

As just suggested [here](https://github.com/Expensify/App/issues/29403#issuecomment-1880902424), i can just (temporarily) create somewhat of a `NoopProvider` or adapt the `MemoryOnlyProvider` to noop. Though in follow-up PR i'd like to replace `OnyxCache`

Reply:
> > completely removing OnyxCache in favor of MemoryOnlyProvider in a follow-up PR, since it's essentially doing the same thing
> 
> Just my opinion, but possibly sounds like more work for unclear gain.

I think the gain is definitely a less complex implementation of `Onyx.js` and the ability to add additional fallback storage providers, and/or just add more storage providers.

Reply:
I see the noop provider was added, but I don't see it referenced anywhere. Maybe I misunderstood the purpose of it?

Reply:
@chrispader you might have missed this, but I'm still waiting for an answer to my previous question here to conclude my review of this PR.

Reply:
Bump @chrispader I'm still waiting for an answer on this.

Reply:
@tgolen sorry for the delay.

I added the `NoopProvider` just to show what i meant for now. I'm going to simplify this PR and remove and unnecessary overhead for now and just use the `NoopProvider` instead of actually using a working fallback provider.

Based on what @marcaaron suggested in https://github.com/Expensify/App/issues/29403#issuecomment-1881775744